### PR TITLE
[FIX JENKINS-48407] Permission issue after upgrade to 2.93

### PR DESCRIPTION
**Pull request by batmat**
See [JENKINS-48407](https://issues.jenkins-ci.org/browse/JENKINS-48407).

This was caused by https://github.com/jenkinsci/jenkins/pull/2548

Basically, NIO's `Files.createTempFile()` creates file with `0600` instead of `0644` previously with pre-NIO.
So the path chosen is the following: instead of applying `0644` back blindly:
* if the file exists, we re-apply the existing permissions
* if it does not, we use the default system permissions, not computed once at Jenkins startup.

### Proposed changelog entries

* [JENKINS-48407](https://issues.jenkins-ci.org/browse/JENKINS-48407) : Fix permission issue in 2.93 caused by https://github.com/jenkinsci/jenkins/pull/2548

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees

### CODE REVIEW

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -5117,6 +5123,47 @@ private static void computeVersion(ServletContext context) {
...
+        int maxAttempts = 10;
+        Set<PosixFilePermission> posixFilePermissions = null;
+        while (maxAttempts > 0) {
+            maxAttempts--;
+            try {
+                Path newFile = null;
+                try {
+                    newFile = Files.createFile(
+                            Paths.get("permission-getter-random-file" + RandomUtils.nextLong()));
+                    posixFilePermissions = Files.getPosixFilePermissions(newFile);
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: throws `UnsupportedOperationException` if the associated file system does not support the PosixFileAttributeView

This needs to be caught.

Also if I am not mistaken this creates the file in the CWD of the process (which may not be writable - esecially for a service) and not in JENKINS_HOME (which may even be a different filesystem with a different permission system entirly).

File: core/src/main/java/hudson/util/AtomicFileWriter.java [hunk 2]
```diff
@@ -111,7 +116,12 @@ public AtomicFileWriter(@Nonnull Path destinationPath, @Nonnull Charset charset)
         }
 
         try {
-            tmpPath = Files.createTempFile(dir, "atomic", "tmp");
+            Set<PosixFilePermission> permissions = Jenkins.DEFAULT_FILE_PERMISSIONS;
+            if (Files.exists(destinationPath)) {
+                permissions = Files.getPosixFilePermissions(destinationPath);
+            }
+            tmpPath = Files.createTempFile(dir, "atomic", "tmp", PosixFilePermissions.asFileAttribute(permissions));
```
Comment (jtnord) [inline core/src/main/java/hudson/util/AtomicFileWriter.java]: javadoc does not list `PosixFilePermissions.asFileAttribute` as null safe not document what happens if null is passed in.

File: core/src/main/java/hudson/util/AtomicFileWriter.java [hunk 3]
```diff
@@ -111,7 +116,12 @@ public AtomicFileWriter(@Nonnull Path destinationPath, @Nonnull Charset charset)
         }
 
         try {
-            tmpPath = Files.createTempFile(dir, "atomic", "tmp");
+            Set<PosixFilePermission> permissions = Jenkins.DEFAULT_FILE_PERMISSIONS;
+            if (Files.exists(destinationPath)) {
```
Comment (jtnord) [inline core/src/main/java/hudson/util/AtomicFileWriter.java]: Do not bother doing this.
AtomicFileWriter has no way to specify previously and would have kept the OS default.
If anything then ATF API should be expanded to take a permission set but this is out of scope.

Also this will fail to run on some file systems causing exceptions that are not caught to always be thrown - and hence no file could ever be written

File: core/src/test/java/hudson/util/AtomicFileWriterTest.java [hunk 4]
```diff
@@ -124,4 +142,72 @@ public void badPath() throws Exception {
                        containsString("exists and is neither a directory nor a symlink to a directory"));
         }
     }
+
+    @Issue("JENKINS-48407")
+    @Test
+    public void preserveSpecificPermissions() throws IOException, InterruptedException {
```
Comment (jtnord) [inline core/src/test/java/hudson/util/AtomicFileWriterTest.java]: this is actually changing existing behaviour is it not.  Before the permissions would have been whatever we created the new file as (as it was create new file, delete old file, rename new file).

I think this is possibly bad behaviour and we should not try and do this.

File: core/src/test/java/hudson/util/AtomicFileWriterTest.java [hunk 5]
```diff
@@ -124,4 +142,72 @@ public void badPath() throws Exception {
...
+        // when
+        AtomicFileWriter w = new AtomicFileWriter(newFilePath, StandardCharsets.UTF_8);
+        w.write("whatever");
+        w.commit();
+
+        // then
+        assertFalse(w.getTemporaryPath().toFile().exists());
+        assertTrue(newFilePath.toFile().exists());
+
+        final Set<PosixFilePermission> posixFilePermissions = Files.getPosixFilePermissions(newFilePath);
```
Comment (jtnord) [inline core/src/test/java/hudson/util/AtomicFileWriterTest.java]: posix on a non posix system will blow up with an exception that is unhandled.

https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#getPosixFilePermissions-java.nio.file.Path-java.nio.file.LinkOption...-

> Throws:
UnsupportedOperationException - if the associated file system does not support the PosixFileAttributeView

**Review by jtnord — CHANGES_REQUESTED** [review]
barfs if tests or Jenkins runs on a non poisx file system.

Changes behaviour over the previous implementation)

Not convinced the logic to get the default permission is correct either - as it looks like it uses the process CWD which may not be writable.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java]:
```
[ERROR] validateAntFileMaskBounded(hudson.FilePathTest)  Time elapsed: 1.431 s  <<< ERROR!
java.lang.ExceptionInInitializerError
	at hudson.ExtensionList.lookup(ExtensionList.java:418)
	at hudson.FilePath.act(FilePath.java:984)
	at hudson.FilePath.act(FilePath.java:976)
	at hudson.FilePath.mkdirs(FilePath.java:1159)
	at hudson.FilePathTest.validateAntFileMaskBounded(FilePathTest.java:553)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:236)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:386)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:323)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:143)
Caused by: java.lang.UnsupportedOperationException
	at sun.nio.fs.WindowsFileSystemProvider.readAttributes(WindowsFileSystemProvider.java:192)
	at java.nio.file.Files.readAttributes(Files.java:1737)
	at java.nio.file.Files.getPosixFilePermissions(Files.java:2004)
	at jenkins.model.Jenkins.<clinit>(Jenkins.java:5145)
	... 32 more

```

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 6]
```diff
@@ -5117,6 +5123,47 @@ private static void computeVersion(ServletContext context) {
...
 
+    /**
+     * The default set of permissions for a new file on the platform (computed once at Jenkins startup).
+     * (provides for instance an indirect access to <code>umask</code> for Unices).
+     */
+    @Nullable
+    public static final Set<PosixFilePermission> DEFAULT_FILE_PERMISSIONS;
+
+    static {
+        LOGGER.log(Level.INFO, "Computing default file permissions");
```
Comment (dwnusbaum) [inline core/src/main/java/jenkins/model/Jenkins.java]: I'm not convinced that this is necessary. `Files#createFile` calls `Files#newByteChannel`,
and [UnixFileSystemProvider#newByteChannel](http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/d5a00b1e8f78/src/solaris/classes/sun/nio/fs/UnixFileSystemProvider.java#l203) uses [UnixFileModeAttribute#ALL_READWRITE](http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/d5a00b1e8f78/src/solaris/classes/sun/nio/fs/UnixFileModeAttribute.java#l37) (666) on the new file by default. `Files#createTempFile` calls `TempFileHelper#create` which calls `File#createFile` and then `Files#newByteChannel` but it passes permissions 600 which are used instead of the default. We could just past 666 permissions to `Files#createTempFile` and the code path would be the same (for java versions built from openjdk).

Maybe there is a valid reason to do the default check as it is here if we are concerned that another supported java implementation does something differently? The `Setting Initial Permissions` section of the documentation for [PosixFileAttributeView](https://docs.oracle.com/javase/7/docs/api/java/nio/file/attribute/PosixFileAttributeView.html) mentions that umask might apply to newly created files, but I don't know if that is enough documentation to be confident in the behavior.

File: core/src/test/java/hudson/util/AtomicFileWriterTest.java (hunk 4 above)
Comment (dwnusbaum) [reply core/src/test/java/hudson/util/AtomicFileWriterTest.java]: I initially thought preserving permissions would be a good idea but now I agree with @jtnord. I believe this would also cause files who got new permissions after #2548 to keep the same bad permissions for anyone who is running the latest weekly builds.

Comment (batmat) [reply core/src/test/java/hudson/util/AtomicFileWriterTest.java]: I agree I'm not keeping the following behaviour, but it was on purpose as explained in [my comment](https://issues.jenkins-ci.org/browse/JENKINS-48407?focusedCommentId=322043&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-322043).

So, though I obviously agree this is not preserving the previous behaviour, I still think this is the right thing to do:
* preserve the previous perms, if the file was already existing
* use the default permissions, iow respecting `umask` (if we can, sigh, does not seem so easy from Java code...), and respecting **for that case** the previous behaviour.

Now, from reading below and so on, does not look like it'd be so easy, especially as on Windows that permission retrieval seems to barf... WAT.

**Comment by oleg-nenashev** [comment]
@batmat any plans to work on it?

**Comment by batmat** [comment]
@oleg-nenashev yes, definitely. I hope to be able to work on it this week, or next one. In any case, as said in the JIRA, I think commits for JENKINS-34855 should only be picked up for next LTS iff this one is fixed too. For LTS stability.

**Comment by batmat** [comment]
Superseded by the far simpler #3233
